### PR TITLE
refactor: better architecture for differ library

### LINKED ISSUE #36: refactor: better architecture for differ library
With #10 open (for computing scores), and #11 (for computing weighted scores), I'd like to propose something that I think would be more scalable than the current design of the library. It still similarly models the current way of how a developer would go about implementing `StringDiffAlgorithm`, but ultimately the trait gets remodeled into a new struct, `Diff`, which allows avoid calling helper methods e.g `get_operations_matrix()` multiple times. 

Some notes:
 - This should also help make implementing #16 scalable in a way that code does not have to be duplicated.
 - `Diff` and `DiffScoreConfig` intentionally do not have the `String` prefix unlike the current types for the case where we allow computing different data types besides just strings (with generic types). What we can do though is rename the current types to remove the `String` prefix, and then introduce these types.

## DiffScoreConfig
The `DiffScoreConfig` is a type of structure that only needs to get created once, and then passed around as a reference to avoid copying (since the size of a struct will be quite big initially and will ). This type implements the `Default` trait, to provide sensible default weighted values for different type of operations. 

An alternative is to use a `HashMap`, but this would involve creating a key type for `K` like another enum (since `StringDiffOpKind` is specifically for storing the values for each operation), as well as void re-hashing and re-allocations (`HashMap` is a DST). By having specific properties instead, the `DiffScoreConfig` struct will have a guaranteed fized-size at compile-time.

```rs
pub struct DiffScoreConfig {
   pub sub_cost: f32,
   pub indel_cost: f32,
   pub transpose_cost: f32,
   // future properties here as needed
}

impl Default for DiffScoreConfig {
   fn default() -> Self { /* ... */ }
}

// used like:
let mut score_config = DiffScoreConfig::default();
score_config.sub_cost = 0.75f32;
score_config.indel_cost = 1.0f32;

// or, if the defaults is good enough for the user,
// it doesn't need to be mutable
let score_config = DiffScoreConfig::default();
```

## Diff
The `Diff` struct would be the return type given by diffing algorithms. It has an `ops` property which holds a slice (`&[T]`), which are immutable pointers. It also holds a `total_len`, to know how to compute the score. 

It does *not* however store an instance of a `DiffScoreConfig`; that's because a `Diff` type will be returned by the hamming + levenshtein distance algorithms (and future algorithms). This would mean require having to pass in `DiffScoreConfig` as an argument for everything, and because `DiffScoreConfig` is designed to be mutable.

The `similarity()` and `difference()` are methods instead of properties since they would take time to compute, The job of a constructor is only to initialize the state/fields of a type. 

The jobs of the distance algorithms are *purely* to compute the list of difference operations, and the user won't *always* need to know what the similarity and difference scores are.

```rs
pub struct Diff {
   pub ops: &[StringDiffOp],
   pub total_len: usize,
}

impl Diff {
    pub fn new(diffs: &[StringDiffOp], total_len: usize) -> Self { Self {} }
    pub fn distance(&self) -> usize {
       self.diffs.len()
    }

    pub fn similarity(&self, score: &DiffScoreConfig) -> f32 { /* ... */ }
    pub fn difference(&self, score: &DiffScoreConfig) -> f32 { /* ... */ }
}
```

## Algorithms
This leaves the `StringDiffAlgorithm`, and the hamming distance and levenshtein. Further thinking, I think it would actually make more sense to remove the `StringDiffAlgorithm` trait, and turn both hamming + levenshtein algorithms from structs into pure functions. Helper methods can still be only available internally (`pub(crate)`). This would leave the signatures of the public API to look like the following. Of course, both `hamming.rs` and `levenshtein.rs` would still be separate.

```rs
// hamming.rs
pub fn hamming<'a>(&self, s1: &'a str, s2: &'a str) -> Diff {}

// levenshtein.rs
pub fn levenshtein<'a>(&self, s1: &'a str, s2: &'a str) -> Diff {}
```

### PR DESCRIPTION
**Pull request by notalfredo**
fixes #36

### CODE REVIEW

File: crates/differ/README.md [hunk 1]
```diff
@@ -16,82 +16,73 @@ differ-rs = "0.0.0"
...
+use differ_rs::levenshtein;
 
 fn main(){
-    let my_levensthein = LevenshteinDistance {};
+	let levensthein_edit_distance = levenshtein("Sitting", "Kitten").distance();
 
-    let edit_distance = my_levensthein.distance("Sitting", "Kitten");
-    
-    assert_eq!(3, edit_distance)
+    assert_eq!(3, levensthein_edit_distance);
```
Comment (neoncitylights) [inline crates/differ/README.md]: Can you make the whitespace consistent here?

File: crates/differ/src/apply_diff.rs [hunk 2]
```diff
@@ -11,7 +11,7 @@ pub(crate) fn remove(start: usize, stop: usize, s: &str) -> String {
 	result
 }
 
-pub fn apply_diff(s: &str, diffs: Vec<StringDiffOp>) -> String {
+pub fn apply_diff(s: &str, diffs: &Box<[StringDiffOp]>) -> String {
```
Comment (neoncitylights) [inline crates/differ/src/apply_diff.rs]: apply_diff() can and still should accept Vec<StringDiffOp>, it's specifically for the `Diff` type

File: crates/differ/src/diff_score.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+pub struct DiffScoreConfig {
```
Comment (neoncitylights) [inline crates/differ/src/diff_score.rs]: Can you take out `DiffScoreConfig` until we implement it for #10?

File: crates/differ/src/diff.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+use crate::StringDiffOp;
```
Comment (neoncitylights) [inline crates/differ/src/diff.rs]: Since this type is one of the foundational types, it should also exist in lib.rs

File: crates/differ/src/diff.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+#[derive(Debug, PartialEq)]
```
Comment (neoncitylights) [inline crates/differ/src/diff.rs]: This should also derive the `Eq` trait

File: crates/differ/src/diff.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+	pub fn distance(&self) -> usize {
+		self.ops.len()
+	}
+
+	pub fn operations(&self) {
+		for i in self.ops.iter() {
+			println!("{:?}", i);
+		}
+	}
```
Comment (neoncitylights) [inline crates/differ/src/diff.rs]: This method is out of scope for the `Diff` type since it's IO and should be removed; it's up to the library user to define how they want to debug and print out information with their data types

File: crates/differ/src/lib.rs [hunk 7]
```diff
@@ -39,7 +42,50 @@ impl StringDiffOp {
...
-	fn diff<'a>(&self, s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
-	fn distance<'a>(&self, s1: &'a str, s2: &'a str) -> usize;
+pub(crate) fn get_operation_matrix(
+	s1: &str,
+	s2: &str,
+	dist_with_dir: fn(isize, isize, isize) -> (isize, char),
+	init_vec: fn(&mut Vec<Vec<isize>>, usize, usize),
+	char_match: isize,
+	not_char_match: isize,
+	indent_cost: isize,
```
Comment (neoncitylights) [inline crates/differ/src/lib.rs:67]: Why are these `isize`?

File: crates/differ/src/lib.rs [hunk 8]
```diff
@@ -39,7 +42,50 @@ impl StringDiffOp {
...
+	for j in 1..second_string_len + 1 {
+		dir_vector[j][0] = '^';
+	}
+	for i in 1..first_string_len + 1 {
+		dir_vector[0][i] = '<';
+	}
+
+	for i in 1..second_string_len + 1 {
+		for j in 1..first_string_len + 1 {
+			let diagnal_gap_cost: isize;
```
Comment (neoncitylights) [inline crates/differ/src/lib.rs]: `diagonal_gap_cost`

File: crates/differ/README.md [hunk 9]
```diff
@@ -16,82 +16,73 @@ differ-rs = "0.0.0"
 ```
 
 ## Features
-* `apply_diff`: Allows users to apply delta vectors in order to transform a words.
-* `extra_traits`: all `struct`s implemented in `differ-rs` are `HammingDistance` and `LevenshteinDistance`. Each Struct implements the `diff` and `distance` methods. 
+* `apply_diff` function: Allows users to apply deltas in order to transform a words.
+* `Diff` struct: Contains a Box<> of operations between two strings. Also keeps track of length of longest string. Has methods that allows users to get the edit distance between two words, and view delta operations. 
+* `levenshtein` function: Returns a Diff struct between string 1 and string 2 using levenshtein algorithm. 
+* `hamming` function: Returns a Diff struct between string 1 and string 2 hamming algorithm.
```
Comment (neoncitylights) [inline crates/differ/README.md]: It's obvious that these are an algorithm, but the reader might not know what the algorithm does specifically. Can you go into detail on these? E.g, explain that the Levenshtein algorithm can detect insertions, deletions, and substitutions, while the Hamming algorithm can only detect substitutions. 

Also make sure to capitalize proper nouns when they're within the sentences (hamming should be Hamming). Also, to indicate that they are functions, it should be written with parentheses, like: `hamming()` function: [...]

File: crates/differ/README.md [hunk 10]
```diff
@@ -16,82 +16,73 @@ differ-rs = "0.0.0"
...
+* `apply_diff` function: Allows users to apply deltas in order to transform a words.
+* `Diff` struct: Contains a Box<> of operations between two strings. Also keeps track of length of longest string. Has methods that allows users to get the edit distance between two words, and view delta operations. 
+* `levenshtein` function: Returns a Diff struct between string 1 and string 2 using levenshtein algorithm. 
+* `hamming` function: Returns a Diff struct between string 1 and string 2 hamming algorithm. 
 
 ## How it works
 * `apply_diff` works by giving a string and a transformation vector to the method. Then the transformation vector is applied to the string given in the first argument.
-* `StringDiffAlgorithm` provides two methods `diff` which gives you a transformation vector from the first to second string. The `distance` method gives you the edit distance from the frist argument to the second argument. The structs `HammingDistance` and `LevenshteinDistance` have their own implementations for each method.
+* `Diff` works by hodling a Box<> of operations, and longest length between any two strings. Both the `levenshtein`, `hamming` algorithm return this struct.
```
Comment (neoncitylights) [inline crates/differ/README.md]: - typo: `holding`
 - Use rust syntax here, it should be `Box<StringDiffOp>`; `Box<>` is an invalid type
 - This can be written more succinctly, like:
 ```
 Diff holds a `Box<StringDiffOp>`, and the longest length of any two strings.
 ```
 - Second sentence needs an "and": Both the `levenshtein()` and `hamming()` functions return this struct.

File: crates/differ/README.md [hunk 11]
```diff
@@ -16,82 +16,73 @@ differ-rs = "0.0.0"
...
 fn main(){
-    let my_hamming = HammingDistance {};
+	let kathrin_edit_distance = hamming("karolin", "kathrin");
 
-    let delta_vec = my_hamming.diff("karolin", "kathrin");
-    
-    for i in delta_vec.iter(){
-        println!("{:?}", i);
-    }
+    kathrin_edit_distance.operations();
```
Comment (neoncitylights) [inline crates/differ/README.md]: You can put back the IO code here by accessing the `ops` property:
```rs
for diff_op in delta_vec.ops {
    println!("{:?}", i);
}
```

File: crates/differ/src/apply_diff.rs [hunk 12]
```diff
@@ -37,47 +37,50 @@ mod tests {
...
+		assert_eq!(
+			String::from("SETS"),
+			super::apply_diff("RESET", &test_box_3)
+		);
+		assert_eq!(String::from("heeoy"), super::apply_diff("hey", &test_box_4));
 		assert_eq!(
 			String::from("skater"),
-			super::apply_diff("kate", test_vec_5)
+			super::apply_diff("kate", &test_box_5)
 		);
```
Comment (neoncitylights) [inline crates/differ/src/apply_diff.rs]: Can you make the formatting for all of these `assert_eq!()` macro calls consistent, so they're not broken up into multiple lines?

**Review by neoncitylights — CHANGES_REQUESTED** [review]
Thank you! Here's the first initial review.

File: crates/differ/src/lib.rs (hunk 7 above)
Comment (notalfredo) [reply crates/differ/src/lib.rs:67]: Since I wanted to make the function as generic as possible, the bio_crate requires these values to be isize since it uses negative numbers.

File: crates/differ/src/apply_diff.rs (hunk 12 above)
Comment (notalfredo) [reply crates/differ/src/apply_diff.rs]: `cargo fmt` does that

File: crates/differ/README.md [hunk 13]
```diff
@@ -16,81 +16,75 @@ differ-rs = "0.0.0"
...
+* `levenshtein()`: Returns a Diff struct between string 1 and string 2. Levenshtein algorithm can detect insertions, deletions, and substitutions. 
+* `hamming()`: Returns a Diff struct between string 1 and string 2. Hamming algorithm can only detect substitutions, and string 1 and string 2 must me equal length.
+  
 ## How it works
-* `apply_diff` works by giving a string and a transformation vector to the method. Then the transformation vector is applied to the string given in the first argument.
-* `StringDiffAlgorithm` provides two methods `diff` which gives you a transformation vector from the first to second string. The `distance` method gives you the edit distance from the frist argument to the second argument. The structs `HammingDistance` and `LevenshteinDistance` have their own implementations for each method.
+* `apply_diff()` works by giving a string and a transformation vector to the method. Then the transformation vector is applied to the string given in the first argument.
+* `Diff` holds a `Box<StringDiffOp>`, and the longest length of any two strings. Both `levenshtein()`, and `hamming()`  eturn this struct.
```
Comment (neoncitylights) [inline crates/differ/README.md:26]: typo near end of sentence: `return this struct`

File: crates/differ/src/lib.rs [hunk 14]
```diff
@@ -39,7 +38,69 @@ impl StringDiffOp {
...
+pub struct Diff {
+	pub ops: Box<[StringDiffOp]>,
+	pub total_len: usize,
+}
+
+impl Diff {
+	pub fn new(diffs: Vec<StringDiffOp>, total_len: usize) -> Self {
+		Self {
+			ops: diffs.into_boxed_slice(),
+			total_len: total_len,
```
Comment (neoncitylights) [inline crates/differ/src/lib.rs:51]: this can simplify to `total_len`, no need to declare it twice

**Review by neoncitylights — APPROVED** [review]
